cpp/src/rolling/rolling_detail.cuh

```diff
@@ -279,6 +282,79 @@ struct DeviceRollingCountAll {
   }
 };
 
+/**
+ * @brief Operator for applying a VAR rolling aggregation on a single window.
+ */
+template <typename InputType>
+struct DeviceRollingVariance {
+  size_type min_periods;
+  size_type ddof;
+
+  // what operations do we support
+  template <typename T = InputType, aggregation::Kind O = aggregation::VARIANCE>
+  static constexpr bool is_supported()
+  {
+    return is_fixed_width<InputType>() and not is_chrono<InputType>();
+  }
+
+  DeviceRollingVariance(size_type _min_periods, size_type _ddof)
+    : min_periods(_min_periods), ddof{_ddof}
+  {
+  }
+
+  template <typename OutputType, bool has_nulls>
+  bool __device__ operator()(column_device_view const& input,
+                             column_device_view const&,
+                             mutable_column_device_view& output,
+                             size_type start_index,
+                             size_type end_index,
+                             size_type current_index)
+  {
+    using DeviceInputType = device_storage_type_t<InputType>;
+
+    // valid counts in the window
+    cudf::size_type const count =
+      has_nulls ? thrust::count_if(thrust::seq,
+                                   thrust::make_counting_iterator(start_index),
+                                   thrust::make_counting_iterator(end_index),
+                                   [&input](auto i) { return input.is_valid_nocheck(i); })
+                : end_index - start_index;
+
+    // Variance/Std is non-negative, thus ddof should be strictly less than valid counts.
+    bool output_is_valid = (count >= min_periods) and (ddof < count);
+
+    if (output_is_valid) {
+      // Welford algorithm
+      // See https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance
+      OutputType m{0}, m2{0};
+      size_type running_count{0};
+
+      for (size_type i = start_index; i < end_index; i++) {
+        if (has_nulls and input.is_null_nocheck(i)) { continue; }
```

Comment:
What is the expected output when all input rows in the (valid) window are null? It appears that `output_is_valid` is `true`, but we land up not writing anything to `output.element<OutputType>(current_index)`.

Reply:
Hmm, that's a good catch. In pandas if all rows in window is null, the result is NaN. Maybe the corresponding behavior in libcudf is to return null at these rows too?

Reply:
cc @revans2

Reply:
For `StddevPop` or `VariancePop` if the number of valid entries is 0, then the result is null.
For `StddevSamp` or `VarianceSamp` if the number of valid entries is 0, then the result is null, but if the number of valid entries is 1, then the result is NaN.

Not sure in this case for pandas if `NaN` is `null` or `NaN` is `NaN`, but either way I had already asked about this. We can fix it up in post processing if we have to.  A valid count is a fairly fast aggregation to do, so we can always do the exact same post processing that Spark is doing.  That said it would be nice to not have to do it.

Reply:
Pandas uses NaN as default marker for invalid value. However, libcudf doesn't have to match pandas semantics and should be true to itself (as similar in many other situations). To me, what makes sense for these two corner cases:

- If all inputs in the window is NULL, the result should be NULL. Because invalid input cannot lead to a valid output.
- If there's only 1 valid input and `ddof == 1`, technically what happens is `division by 0` problem. Either:
    - Follow existing cudf::reduction::op::var behavior, surface the division by zero result without checks. https://github.com/rapidsai/cudf/blob/8aefeb49bec96e3bc27a05276e66471f8ca7f966/cpp/include/cudf/detail/reduction_operators.cuh#L260-L263
    - Follow IEEE754: `+-inf` for numerator != 0, `NaN` for `numerator == +-0`.

Reply:
if we have to choose one or ther other, we should return null. null can be post-processed to another value, but inf is ambiguous and not safe to post-process.

Reply:
that said, mathematically there is usually an identity value associated with operationa. addition and subtraction has 0, multiplication and division has 1. string concatenation has "" (empty string). is there an appropriate identity value here?

Reply:
+-inf is an in correct answer, because in the case there are no values to divide, no division should take place. the average of nothing is not +-inf, it is undefined.

Reply:
> is there an appropriate identity value here?

I'm not sure how to define an identity value for variance. As shown in: https://en.wikipedia.org/wiki/Identity_element, identity is a special element in the set w.r.t. certain **binary operation**, and since variance is not a binop, maybe there isn't one?

Reply:
> identity is a special element in the set w.r.t. certain binary operation

That's a good point! Defining an identity value would first require defining a type that represents the values of the group the identify belongs to, and expressing variance as a binary operation on that type. Suffice to say, no strictly-numeric type would be a member of that group, and it's highly unlikely that that +-inf would be that identity value, either.

So again, null is probably the best answer here. :)

Reply:
I don't think you want to go down the route of overthinking this mathematically, in this case that's bound to devolve into abstract measure theoretic discussions long before it produces anything practically relevant. I agree with Michael's original proposal, with a slight preference for IEEE conformance to the behavior of 0/0 vs x/0 for any nonzero x. That seems like it would match Spark semantics reasonably well and just require an inf->NaN substitution if I understood Bobby's comment correctly.

Reply:
We should not yield to spark or pandas on this matter. We should do what is most flexible, and that is returning null. We can easily convert nulls to +/- inf for spark, but we cannot deterministically convert +/- inf to nulls, because if any of the inputs are inf, we encounter ambiguity.

Either we need to return null and let the user adjust the result as needed, or we need to let the user inject the default behavior as a fallback (identity) value.

Reply:
> We can easily convert nulls to +/- inf for spark

That is conditionally true when there is only a single cause that results in `null` (divide by zero). In this PR, there are other reason that may lead to a null:
- all inputs of the window is null
- the valid inputs of the window is less than `min_periods` parameter.

We cannot uniquely determine the cause of nulls in practice without inevitably introspecting the input.

Reply:
I'm not suggesting that we match Spark semantics, I was saying that this choice is compatible because Spark could unambiguously parse those outputs. I see that input infs make this ambiguous if users actually need to distinguish those cases, although I'd argue that they probably shouldn't be. Moreover, nulls do not fix this problem here because nulls can _also_ be generated in multiple different ways (for example, how would you differentiate between the 0 and 1 non-null element window cases?). No matter what choice we make, some amount of data introspection will be necessary to determine why a particular output is an exceptional (NaN, null, or inf) value. 

I recognize the need to prioritize flexibility over most other considerations to enable all the strange semantics of different libcudf consumers, but I'm not really convinced that always choosing null is any more helpful than the alternatives here. That being the case I think we're better off providing the most mathematically precise result we can generate.

Reply:
I guess Github doesn't refresh threads automatically :) My example of where the all nulls approach fails is the same as Michael's first one.

Reply:
How difficult would it be to allow the user to specify the default value for the exceptional cases?

Reply:
Not particularly hard, we can augment `var_std_aggregation` to take extra parameters to let user specify these default values.

Reply:
My main concern with that approach would be consistency. Would we need to modify all relevant APIs (variance/std for non-rolling cases, mean for a set of nulls, etc) to provide those parameters? If we choose to distinguish the 0 nulls case from the 1 null, ddof=1 case for variance, would that apply analogously for higher order moments? For instance, 2 nulls, ddof=2 should be the same for skewness.

Reply:
lets go with whatever is most flexible for now, then, which to @vyasr’s point about exceptional cases and the likelihood if inf input values seems to be…

return null for exceptional cases, and return +-inf for the div-by-zero case. anything more, and we need to reexamine the aggregation APIs as a whole.